### PR TITLE
Client (mesh egress) discovery for Jobs and Numaflow (vertices and monovertices)

### PR DESCRIPTION
**Pull request by aattuluri**
### Checklist
🚨 Please review this repository's [contribution guidelines](./CONTRIBUTING.md).

- [ ] I've read and agree to the project's contribution guidelines.
- [ ] I'm requesting to **pull a topic/feature/bugfix branch**.
- [ ] I checked that my code additions will pass code linting checks and unit tests.
- [ ] I updated unit and integration tests (if applicable).
- [ ] I'm ready to notify the team of this contribution.

### Description
What does this change do and why?
This change introduces client discovery where there is no ingress but only mesh egress required like k8s Jobs, numaflow Vertices and Monovertices.

[Link to related ISSUE]

Thank you!

### CODE REVIEW

File: admiral/pkg/controller/admiral/monovertex.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+func (d *MonoVertexController) Updated(ctx context.Context, obj interface{}, oldObj interface{}) error {
+	//Not Required, to be handled via off boarding
+	return nil
+}
+
+func addUpdateMonoVertex(j *MonoVertexController, ctx context.Context, obj interface{}) error {
+	monoVertex, ok := obj.(v1alpha1.MonoVertex)
+	if !ok {
+		return nil
```
Comment (kpharasi) [inline admiral/pkg/controller/admiral/monovertex.go:211]: return an error if !ok

File: admiral/pkg/controller/admiral/monovertex.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+func NewMonoVertexCache() *monoVertexCache {
+	return &monoVertexCache{
+		cache: make(map[string]*MonoVertexEntry),
+		mutex: &sync.Mutex{},
+	}
+}
+
+
+func (p *monoVertexCache) getK8sObjectFromMonoVertex(monoVertex v1alpha1.MonoVertex) *K8sObject{
```
Comment (kpharasi) [inline admiral/pkg/controller/admiral/monovertex.go:49]: We probably don't need a pointer receiver here

File: admiral/pkg/client/loader/kube_loader.go [hunk 3]
```diff
@@ -85,6 +86,19 @@ func (loader *KubeClientLoader) LoadKubeClientFromConfig(config *rest.Config) (k
 	return kubernetes.NewForConfig(config)
 }
 
+func (loader *KubeClientLoader) LoadNumaflowClientFromPath(kubeConfigPath string) (numaflow.Interface, error) {
```
Comment (kpharasi) [inline admiral/pkg/client/loader/kube_loader.go:89]: should we have this in a new file called - `numaflow_loader.go`?

File: admiral/pkg/clusters/handler.go [hunk 4]
```diff
@@ -31,6 +32,14 @@ type WeightedService struct {
 	Service *coreV1.Service
 }
 
+type K8sObjectMetadata struct {
```
Comment (kpharasi) [inline admiral/pkg/clusters/handler.go:35]: Should we move this to: https://github.com/istio-ecosystem/admiral/blob/585c2ea5d9dab5642352f2c8f14794cb1b8006bb/admiral/pkg/controller/common/types.go?

File: admiral/pkg/controller/admiral/vertex.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+func (d *VertexController) Updated(ctx context.Context, obj interface{}, oldObj interface{}) error {
+	//Not Required, to be handled via off boarding
+	return nil
+}
+
+func addUpdateVertex(j *VertexController, ctx context.Context, obj interface{}) error {
+	vertex, ok := obj.(v1alpha1.Vertex)
+	if !ok {
+		return nil
```
Comment (kpharasi) [inline admiral/pkg/controller/admiral/vertex.go:211]: return an error if !ok

File: admiral/pkg/controller/admiral/vertex.go [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+func NewVertexCache() *vertexCache {
+	return &vertexCache{
+		cache: make(map[string]*VertexEntry),
+		mutex: &sync.Mutex{},
+	}
+}
+
+
+func (p *vertexCache) getK8sObjectFromVertex(vertex v1alpha1.Vertex) *K8sObject{
```
Comment (kpharasi) [inline admiral/pkg/controller/admiral/vertex.go:49]: We probably don't need a pointer receiver here

File: admiral/pkg/controller/admiral/job.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+	//Not Required, to be handled via off boarding
+	return nil
+}
+
+func addUpdateJob(j *JobController, ctx context.Context, obj interface{}) error {
+	job, ok := obj.(*v12.Job)
+	if !ok {
+		return nil
+	}
+	if !common.ShouldIgnore(job.Annotations, job.Labels) {
```
Comment (kpharasi) [inline admiral/pkg/controller/admiral/job.go:214]: Jobs created by cronjobs would also start getting watched by this controller. In the `ShouldIgnore` function we should also ignore jobs which have an `ownerReferences` section. WDYT?

File: admiral/pkg/client/loader/kube_loader.go (hunk 3 above)
Comment (aattuluri) [reply admiral/pkg/client/loader/kube_loader.go:89]: I see istio client loaded in the same file, do you see an advantage in separating it out?

File: admiral/pkg/clusters/handler.go (hunk 4 above)
Comment (aattuluri) [reply admiral/pkg/clusters/handler.go:35]: Good idea, let me do that.

File: admiral/pkg/controller/admiral/controller.go [hunk 8]
```diff
@@ -80,6 +80,19 @@ type Controller struct {
 	informer  cache.SharedIndexInformer
 }
 
+type K8sObject struct {
```
[the remark this answers is not in the document]
Comment (aattuluri) [reply admiral/pkg/controller/admiral/controller.go]: Done

File: admiral/pkg/clusters/handler.go (hunk 4 above)
Comment (aattuluri) [reply admiral/pkg/clusters/handler.go]: Done

File: admiral/pkg/controller/admiral/job.go (hunk 7 above)
Comment (aattuluri) [reply admiral/pkg/controller/admiral/job.go:214]: All mesh annotations go under job spec in a cron job, so it should be fine to support this at the job spec/job level.

File: admiral/pkg/controller/common/common.go [hunk 9]
```diff
@@ -703,6 +707,56 @@ func GetMeshPortsHelper(meshPorts string, destService *k8sV1.Service, clusterNam
...
+
+	//if sidecar not injected
+	if annotations[labelSet.DeploymentAnnotation] != "true" { //Not sidecar injected, we don't want to inject
+		return true
+	}
+
+	if annotations[AdmiralIgnoreAnnotation] == "true" {
+		return true
+	}
+
```
[the remark this answers is not in the document]
Comment (aattuluri) [reply admiral/pkg/controller/common/common.go:726]: I deliberately chose to not do namespace as thats not in use currently and we should skip that to save some cycles.

File: admiral/pkg/controller/admiral/vertex.go (hunk 6 above)
Comment (aattuluri) [reply admiral/pkg/controller/admiral/vertex.go]: @kpharasi You mean to K8sObject can be returned as value and not reference?

File: admiral/pkg/controller/common/common.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kpharasi) [reply admiral/pkg/controller/common/common.go:726]: Got it, makes sense.

File: admiral/pkg/client/loader/kube_loader.go (hunk 3 above)
Comment (kpharasi) [reply admiral/pkg/client/loader/kube_loader.go:89]: Got it, if it has istio client we can keep this here as well. I thought the file is only for k8s client.

File: admiral/pkg/controller/admiral/vertex.go (hunk 6 above)
Comment (kpharasi) [reply admiral/pkg/controller/admiral/vertex.go]: I meant we can directly call this function rather than using the `vertexCache` object to call it  as we are not using any field from the `vertexCache` struct.

File: admiral/pkg/controller/admiral/job.go (hunk 7 above)
Comment (kpharasi) [reply admiral/pkg/controller/admiral/job.go:214]: Got it, makes sense.

File: admiral/pkg/controller/admiral/job.go [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+func NewJobCache() *jobCache {
+	return &jobCache{
+		cache: make(map[string]*JobEntry),
+		mutex: &sync.Mutex{},
+	}
+}
+
+
+func (p *jobCache) getK8sObjectFromJob(job *v12.Job) *K8sObject{
```
[the remark this answers is not in the document]
Comment (aattuluri) [reply admiral/pkg/controller/admiral/job.go]: Done

File: admiral/pkg/clusters/handler.go [hunk 11]
```diff
@@ -323,3 +325,94 @@ func GetServiceWithSuffixMatch(suffix string, services []*coreV1.Service) string
 	}
 	return ""
 }
+
+func HandleEventForClientDiscovery(ctx context.Context, event admiral.EventType, obj *common.K8sObject,
```
Comment (shriramsharma) [inline admiral/pkg/clusters/handler.go:329]: can this func be moved to `clientdiscovery_handler.go`?

File: admiral/pkg/clusters/handler.go [hunk 12]
```diff
@@ -323,3 +325,94 @@ func GetServiceWithSuffixMatch(suffix string, services []*coreV1.Service) string
...
+		return nil
+	}
+
+	ctx = context.WithValue(ctx, "clusterName", clusterName)
+	ctx = context.WithValue(ctx, "eventResourceType", obj.Type)
+
+	if remoteRegistry.AdmiralCache != nil {
+
+		UpdateIdentityClusterCache(remoteRegistry, globalIdentifier, clusterName)
+
```
Comment (shriramsharma) [inline admiral/pkg/clusters/handler.go:362]: I think we should return an error if `remoteRegistry.AdmiralCache == nil` and not move ahead because if `UpdateIdentityClusterCache` is not called then calling Added of dependency controller might not do anything.

Comment (aattuluri) [reply admiral/pkg/clusters/handler.go:362]: Good catch!

File: admiral/pkg/clusters/handler.go (hunk 11 above)
Comment (aattuluri) [reply admiral/pkg/clusters/handler.go]: Done

File: admiral/pkg/clusters/handler.go [hunk 13]
```diff
@@ -323,3 +325,94 @@ func GetServiceWithSuffixMatch(suffix string, services []*coreV1.Service) string
...
+		return nil
+	}
+	ctxLogger := common.GetCtxLogger(ctx, globalIdentifier, "")
+	if commonUtil.IsAdmiralReadOnly() {
+		ctxLogger.Infof(common.CtxLogFormat, event, "", "", "", "processing skipped as Admiral is in Read-only mode")
+		return nil
+	}
+
+	// Should not return early here for TrafficConfig persona, as cache should build up during warm up time
+	if IsCacheWarmupTime(remoteRegistry) && !common.IsPersonaTrafficConfig() {
```
[the remark this answers is not in the document]
Comment (aattuluri) [reply admiral/pkg/clusters/handler.go]: Done

File: admiral/pkg/clusters/dependency_handler.go [hunk 14]
```diff
@@ -87,15 +87,17 @@ func isIdentityMeshEnabled(identity string, remoteRegistry *RemoteRegistry) bool
...
 	updatedDependency *v1.Dependency, remoteRegistry *RemoteRegistry) ([]string, bool) {
 	updatedDestinations := make([]string, 0)
 	existingDestination := remoteRegistry.AdmiralCache.SourceToDestinations.Get(updatedDependency.Spec.Source)
-
 	var nonMeshEnabledExists bool
 	lookup := make(map[string]bool)
-	for _, dest := range existingDestination {
-		lookup[dest] = true
+	//if this is an update, build a look up table to process only the diff
+	if eventType == admiral.Update {
```
Comment (kpharasi) [inline admiral/pkg/clusters/dependency_handler.go:97]: Why was this added?

File: admiral/pkg/controller/admiral/job.go [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+
+	return &jobController, nil
+}
+
+func (d *JobController) Added(ctx context.Context, obj interface{}) error {
+	return addUpdateJob(d, ctx, obj)
+}
+
+func (d *JobController) Updated(ctx context.Context, obj interface{}, oldObj interface{}) error {
+	//Not Required, to be handled via off boarding
```
Comment (kpharasi) [inline admiral/pkg/controller/admiral/job.go:192]: We just want to skip Update? I think this comment is valid for only delete. Same for the other 2.

Comment (aattuluri) [reply admiral/pkg/controller/admiral/job.go:192]: An Update to the job is kind of not relevant as we just need the Job to be registered as a client when its created.

File: admiral/pkg/clusters/dependency_handler.go (hunk 14 above)
Comment (aattuluri) [reply admiral/pkg/clusters/dependency_handler.go:97]: `Add` event for dependency record was never processed. Meaning all calls to this method would assume dependency record update (get from cache, compare that with updated object). Adding this will make sure that Add event will consider all destinations to be processed.

Comment (kpharasi) [reply admiral/pkg/clusters/dependency_handler.go:97]: Cool, thanks!

**Review by kpharasi — APPROVED** [review]
